### PR TITLE
Added open boundary conditions

### PR DESCRIPTION
**Pull request by torrewk**
Open boundary conditions are now available for simulations, in both RPY and SD. Double precision is required, at the moment, in the case of open boundaries.

### CODE REVIEW

**Review by qubixes — CHANGES_REQUESTED** [review]
That's quite the big PR, lot of work done! I have more or less gone through the code as far as I can understand (which actually is not a lot). If I read the code correctly, you haven't added any tests for the new open boundary conditions. Perhaps you can add a simple test for those?

File: jfsd/config.py [hunk 1]
```diff
@@ -71,12 +71,13 @@ class Initialization(NamedTuple):
...
-                raise ValueError("Please supply the numpy file if using the source_type 'file'.")
-            positions = np.load(numpy_file)
+        #elif self.position_source_type == "file":
+        #    if numpy_file is None:
+        #        raise ValueError("Please supply the numpy file if using the source_type 'file'.")
+        #    positions = np.load(numpy_file)
         else:
-            raise ValueError(f"Unknown source_type {self.position_source_type}")
+            positions = np.load(self.position_source_type)
+            #raise ValueError(f"Unknown source_type {self.position_source_type}")
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/config.py:80]: I wasn't able to load initial particle positions from an input file.npy.

File: jfsd/main.py [hunk 2]
```diff
@@ -112,6 +108,148 @@ def main(
...
+        Strength of hydrodynamic friction
+    h0_friction: (float)
+        Range of hydrodynamic friction
+
+    Returns
+    -------
+        trajectory, stresslet_history, velocities, test_result
+
+    """
+    trajectory = stresslet_history= velocities=test_result=0
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/main.py:125]: Just for the 'if' statement later. Will change to None as suggested.

File: jfsd/main.py [hunk 3]
```diff
@@ -112,6 +108,148 @@ def main(
...
+    if(boundary_flag==1):
+        if((N<2) and (HIs_flag>0)):
+            raise ValueError(
+                "Error: trying to use open boundaries hydrodynamic for a single particle. Select 'brownian' and re-run. Abort!")
+        Lx = Ly = Lz = 999999 #set system size to 'infinite' (does not impact performances)
+        config.update("jax_enable_x64", True)  # enable double precision: needed to resolve long-range hydrodynamics
+        
+    print('jfsd is running on device: ',(xla_bridge.get_backend().platform))
+    
+    # positions = positions.astype(np.float64) #need float64 to be the same dtype of arrays used in jax with double precision
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/main.py:138]: Forgot this I guess.

File: jfsd/main.py [hunk 4]
```diff
@@ -127,61 +265,52 @@ def main(
...
-            space.map_product(displacement))(positions, positions)
+        dR = dR.at[:, 0].set(dt * shear_rate * positions.at[:, 1].get()) # Assuming y:gradient direction, x:background flow direction
+        positions = shift(positions+jnp.array([Lx, Ly, Lz])/2, dR) - jnp.array([Lx, Ly, Lz]) * 0.5 # Apply shift
 
+        # Compute new relative displacements between particles
+        displacements_vector_matrix = (space.map_product(displacement))(positions, positions)
         return positions, displacements_vector_matrix
 
-    start_time = time.time()  # perfomances evaluation
+    # start_time = time.time()  # perfomances evaluation
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/main.py:313]: Same here.

File: jfsd/main.py [hunk 5]
```diff
@@ -221,370 +379,241 @@ def update_positions(
...
+                                                         displacements_vector_matrix.at[nl_ff[0, :], nl_ff[1, :]].get())
             if(exitcode_gmres > 0):
                 raise ValueError(
                     f"GMRES (RFD) did not converge! Iterations are {exitcode_gmres}. Abort!")
 
             # compute the near-field hydrodynamic stresslet from the saddle point solution velocity
             if((stresslet_flag > 0) and ((step % writing_period) == 0)):
-                buffer_stresslet = (
-                    resistance.compute_RSU(
+                buffer_stresslet = (resistance.compute_RSU(
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/main.py:524]: Will clean this too.

File: jfsd/main.py [hunk 6]
```diff
@@ -221,370 +379,241 @@ def update_positions(
...
-        
+
         # add applied forces and conservative (pair potential) forces to right-hand side of system
         saddle_b = appliedForces.sumAppliedForces(N, AppliedForce, AppliedTorques, saddle_b, U,
                                                   indices_i_lub, indices_j_lub, displacements_vector_matrix,
-                                                  U_cutoff,HIs_flag,dt)
-        
+                                                  U_cutoff, 2, dt)
+
         # add (-) the ambient rate of strain to the right-hand side (if full hydrodynamics are switched on)
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/main.py:548]: Seems so.

File: jfsd/main.py [hunk 7]
```diff
@@ -221,370 +379,241 @@ def update_positions(
...
+                                                                                              wave_bro_nyind[:,0], wave_bro_nyind[:,1], wave_bro_nyind[:,2],
+                                                                                              gridk, random_array_wave, all_indices_x, all_indices_y, all_indices_z,
+                                                                                              gaussian_grid_spacing1, gaussian_grid_spacing2)
                     rs_linvel, rs_angvel_strain, stepnormff, diag_ff = thermal.compute_real_space_slipvelocity(
                         N, m_self, T, dt, int(n_iter_Lanczos_ff),
                         random_array_real, r, indices_i, indices_j,
                         f1, f2, g1, g2, h1, h2, h3)
-                saddle_b = saddle_b.at[:11*N].add(thermal.convert_to_generalized(
-                    N, ws_linvel, rs_linvel, ws_angvel_strain, rs_angvel_strain))
+                    while((stepnormff > 1e-3) and (n_iter_Lanczos_ff < 150)):
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/main.py:579]: Seems a good idea. Will keep this in consideration for future PR, as these 'magic' values are actually still under revision.

File: jfsd/main.py [hunk 8]
```diff
@@ -721,60 +717,641 @@ def update_positions(
...
+                           N, 0, rs_linvel, 0, rs_angvel_strain)-ff)/jnp.linalg.norm(ff),
                        n_iter_Lanczos_nf,
-                       stepnormnf]            
-            
-        
+                       stepnormnf]
+
     return trajectory, stresslet_history, velocities, test_result
+
+def wrap_RPY(
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/main.py:773]: I agree with avoiding duplication as much as possible. However, sometime seems needed to, indeed, improve performance of jit compiled functions. In other particular cases, I deliberately choose to duplicate for the sake of clarity of the underlying mathematical-numerical method.

File: jfsd/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qubixes) [reply jfsd/config.py:80]: Can you explain in more detail why that is? From what I read, it should work the same, you just have to change "position_source_type" to "file"?

[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/config.py:80]: 'numpy_file' is set to 'None' by default and I am not sure if the user can give it as an input.

[the remark this answers is not in the document]
Comment (qubixes) [reply jfsd/config.py:80]: I agree that currently the user can't give it as input, but then we should just fix it?
